File: Professional Image Modifier 2/[C#]-Professional Image Modifier 2/C#/ImageFunctions/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using System.Drawing.Imaging;
using ImageFunctions.Modifications.CornerDetection;
using AForge;
using ImageFunctions.Forms;
using System.Threading.Tasks;
using AForge.Imaging.Filters;
using ImageFunctions.Controls;


namespace ImageFunctions
{
	public partial class Form1 : Form
	{
		FrmProcessing processing = null;

		#region CrossThread Delegation
		private delegate void LogDelegate(string message);
		private delegate void StatusDelegate(string status, bool Show);
		private delegate void ImageCompleteDelegate(List<IntPoint> Corners);
		#endregion

		#region Private Variables
		private string CurrentImage = null;
		#endregion

		#region Controls
		private SusanCornerProperties SusanProperties;
		#endregion

		#region Enums
		private enum Modifications : int
		{
			None,
			CornerDetection
		};

		private enum Methods : int
		{
			none,
			Susan, // Corner Detection
		}
		#endregion


		public Form1()
		{
			InitializeComponent();

		}


		// Display information on the codecs the system is aware of.
		private void btnSystemCodecInformation_Click(object sender, EventArgs e)
		{
			foreach (ImageCodecInfo ici in ImageCodecInfo.GetImageDecoders())
			{
				Log("*************************************************");
				Log("Name: " + ici.CodecName);
				Log("Dll Name: " + ici.DllName);
				Log("Filename Extension(s): " + ici.FilenameExtension);
				//TODO: Work out how to get Flag information.
				//foreach (ImageCodecFlags icf in ici.Flags)
				//{
				//	Log("Codec Flags: " + icf.ToString());
				//}
				Log("Format Description: " + ici.FormatDescription);
				Log("Mime Type: " + ici.MimeType);
				Log("Signature Masks: " + ConvertToString(ici.SignatureMasks));
				Log("Signature Patterns: " + ConvertToString(ici.SignaturePatterns));
				Log("Codec Version: " + ici.Version);
				Log("*************************************************");
			}
		}

		// Convert byte[][] to string
		private string ConvertToString(byte[][] p)
		{
			StringBuilder sb = new StringBuilder();
			foreach (byte[] b in p)
			{
				sb.Append(System.Text.Encoding.Default.GetString(b));
			}
			return sb.ToString();
		}

		// Write out information to the console window at the bottom of the app
		private void Log(string p)
		{
			if (rtbConOut.InvokeRequired)
			{
				LogDelegate d = new LogDelegate(Log);
				this.Invoke(d, new object[] { p });
			}
			else
			{
				rtbConOut.AppendText(p + Environment.NewLine);
				rtbConOut.Focus();
				//	Application.DoEvents();
			}
		}

		// To reset we just reload the image.
		private void resetToolStripMenuItem_Click(object sender, EventArgs e)
		{
			pbImage.Image = Image.FromFile(CurrentImage);
			pbImage.Zoom = 100;
		}

		// Load new Image(s) into the listbox and display the first image
		private void btnLoadNewImage_Click_1(object sender, EventArgs e)
		{
			string firstImage = null;
			DialogResult dr = openFile.ShowDialog();
			if (dr == System.Windows.Forms.DialogResult.OK)
			{
				foreach (string file in openFile.FileNames)
				{
					if (firstImage == null) firstImage = file;
					lbImageList.Items.Add(file);

				}
				pbImage.Image = Image.FromFile(firstImage);
				CurrentImage = firstImage;
				lbModification.Enabled = true;
			}
		}

		// Display the picture that the user has selected
		private void listBox1_SelectedIndexChanged_1(object sender, EventArgs e)
		{
			pbImage.Image = Image.FromFile(lbImageList.SelectedItem.ToString());
			CurrentImage = lbImageList.SelectedItem.ToString();
		}

		// The user has changed the form of modification they want to use.
		private void lbModification_SelectedIndexChanged(object sender, EventArgs e)
		{
			lbMethods.Enabled = true;
			switch (lbModification.SelectedIndex)
			{
				case (int)Modifications.None:
					break;
				case (int)Modifications.CornerDetection:
					PopulateMethods(Modifications.CornerDetection);
					break;
				default:
					break;
			}

		}

		// Populate the Modification Method list box.
		private void PopulateMethods(Modifications modifications)
		{
			lbMethods.Items.Add("Susan Edge Detection"); // index 0
		}

		private void lbMethods_SelectedIndexChanged(object sender, EventArgs e)
		{
			switch (lbMethods.SelectedIndex)
			{
				case (int)Methods.none:
					break;
				case (int)Methods.Susan:

					DoSusan();
					break;
				default:
					break;
			}
		}

		#region Susan Edge Detection

		private void DoSusan()
		{
			SusanProperties = new SusanCornerProperties();
			SusanProperties.Dock = DockStyle.Fill;
			ControlPanel.Controls.Add(SusanProperties);
		}

		private void s_ImageComplete(List<IntPoint> Corners)
		{
			if (processing != null)
			{
				if (processing.InvokeRequired)
				{
					ImageCompleteDelegate d = new ImageCompleteDelegate(s_ImageComplete);
					this.Invoke(d, new object[] { Corners });
				}
				else
				{
					processing.Close();
					processing = null;
				}

			}
			Log("Detection Finished");
			SetStatus(Corners.Count + " corners detected", true);
			DrawCorners(Corners);
		}

		#endregion

		private void DrawCorners(List<IntPoint> Corners)
		{
			// Load image and create everything you need for drawing
			Bitmap image = new Bitmap(pbImage.Image);
			Graphics graphics = Graphics.FromImage(image);
			SolidBrush brush = new SolidBrush(Color.Red);
			Pen pen = new Pen(brush);


			// Visualization: Draw 3x3 boxes around the corners
			foreach (IntPoint corner in Corners)
			{
				graphics.DrawRectangle(pen, corner.X - 1, corner.Y - 1, 3, 3);
			}

			// Display
			pbImage.Image = image;
		}

		private void SetStatus(string message, bool Show = true)
		{

			if (Show)
			{
				if (statusStrip1.InvokeRequired)
				{
					StatusDelegate d = new StatusDelegate(SetStatus);
					this.Invoke(d, new object[] { message, Show });
				}
				else
				{
					lblStatus.Text = message;

				}

			}
			else
			{
				if (statusStrip1.InvokeRequired)
				{
					StatusDelegate d = new StatusDelegate(SetStatus);
					this.Invoke(d, new object[] { message, Show });
				}
				else
				{
					lblStatus.Text = message;

				}
			}
		}

		private void btnReset_Click(object sender, EventArgs e)
		{
			switch (lbMethods.SelectedIndex)
			{
				case (int)Methods.none:
					break;
				case (int)Methods.Susan:
					SusanCornerProperties sp = ControlPanel.Controls[0] as SusanCornerProperties;
					sp.SetDefaults();
				
					break;
				default:
					break;
			}
		}

		private void btnProcess_Click(object sender, EventArgs e)
		{
			switch (lbMethods.SelectedIndex)
			{
				case (int)Methods.none:
					break;
				case (int)Methods.Susan:
					SusanCornerProperties sp = ControlPanel.Controls[0] as SusanCornerProperties;
					
					processing = new FrmProcessing("Conducting Susan Edge Detection");
					processing.Show();
					Log("Conducting Susan Edge Detection");
					SetStatus("Please wait for edge detection");
					Susan s = new Susan(CurrentImage);
					s.ImageComplete += s_ImageComplete;
					Task t = new Task(() => s.GetCorners(sp.DifferenceThreshold,sp.GeometricalThreshold));
					t.Start();
					
					break;
				default:
					break;
			}
		}



	}
}

File: Professional Image Modifier 2/[VB]-Professional Image Modifier 2/VB/ImageFunctions/Form1.vb
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Linq
Imports System.Text
Imports System.Windows.Forms
Imports System.Drawing.Imaging
Imports ImageFunctions.Modifications.CornerDetection
Imports AForge
Imports ImageFunctions.Forms
Imports System.Threading.Tasks
Imports AForge.Imaging.Filters
Imports ImageFunctions.Controls


Namespace ImageFunctions
	Partial Public Class Form1
		Inherits Form

		Private processing As FrmProcessing = Nothing

		#Region "CrossThread Delegation"
		Private Delegate Sub LogDelegate(ByVal message As String)
		Private Delegate Sub StatusDelegate(ByVal status As String, ByVal Show As Boolean)
		Private Delegate Sub ImageCompleteDelegate(ByVal Corners As List(Of IntPoint))
		#End Region

		#Region "Private Variables"
		Private CurrentImage As String = Nothing
		#End Region

		#Region "Controls"
		Private SusanProperties As SusanCornerProperties
		#End Region

		#Region "Enums"
		Private Enum Modifications As Integer
			None
			CornerDetection
		End Enum

		Private Enum Methods As Integer
			none
			Susan ' Corner Detection
		End Enum
		#End Region


		Public Sub New()
			InitializeComponent()

		End Sub


		' Display information on the codecs the system is aware of.
		Private Sub btnSystemCodecInformation_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btnSystemCodecInformation.Click
			For Each ici As ImageCodecInfo In ImageCodecInfo.GetImageDecoders()
				Log("*************************************************")
				Log("Name: " & ici.CodecName)
				Log("Dll Name: " & ici.DllName)
				Log("Filename Extension(s): " & ici.FilenameExtension)
				'TODO: Work out how to get Flag information.
				'foreach (ImageCodecFlags icf in ici.Flags)
				'{
				'	Log("Codec Flags: " + icf.ToString());
				'}
				Log("Format Description: " & ici.FormatDescription)
				Log("Mime Type: " & ici.MimeType)
				Log("Signature Masks: " & ConvertToString(ici.SignatureMasks))
				Log("Signature Patterns: " & ConvertToString(ici.SignaturePatterns))
				Log("Codec Version: " & ici.Version)
				Log("*************************************************")
			Next ici
		End Sub

		' Convert byte[][] to string
		Private Function ConvertToString(ByVal p()() As Byte) As String
			Dim sb As New StringBuilder()
			For Each b As Byte() In p
				sb.Append(System.Text.Encoding.Default.GetString(b))
			Next b
			Return sb.ToString()
		End Function

		' Write out information to the console window at the bottom of the app
		Private Sub Log(ByVal p As String)
			If rtbConOut.InvokeRequired Then
				Dim d As New LogDelegate(AddressOf Log)
				Me.Invoke(d, New Object() { p })
			Else
				rtbConOut.AppendText(p & Environment.NewLine)
				rtbConOut.Focus()
				'	Application.DoEvents();
			End If
		End Sub

		' To reset we just reload the image.
		Private Sub resetToolStripMenuItem_Click(ByVal sender As Object, ByVal e As EventArgs) Handles resetToolStripMenuItem.Click
			pbImage.Image = Image.FromFile(CurrentImage)
			pbImage.Zoom = 100
		End Sub

		' Load new Image(s) into the listbox and display the first image
		Private Sub btnLoadNewImage_Click_1(ByVal sender As Object, ByVal e As EventArgs) Handles btnLoadNewImage.Click
			Dim firstImage As String = Nothing
			Dim dr As DialogResult = openFile.ShowDialog()
			If dr = System.Windows.Forms.DialogResult.OK Then
				For Each file As String In openFile.FileNames
					If firstImage Is Nothing Then
						firstImage = file
					End If
					lbImageList.Items.Add(file)

				Next file
				pbImage.Image = Image.FromFile(firstImage)
				CurrentImage = firstImage
				lbModification.Enabled = True
			End If
		End Sub

		' Display the picture that the user has selected
		Private Sub listBox1_SelectedIndexChanged_1(ByVal sender As Object, ByVal e As EventArgs) Handles lbImageList.SelectedIndexChanged
			pbImage.Image = Image.FromFile(lbImageList.SelectedItem.ToString())
			CurrentImage = lbImageList.SelectedItem.ToString()
		End Sub

		' The user has changed the form of modification they want to use.
		Private Sub lbModification_SelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs) Handles lbModification.SelectedIndexChanged
			lbMethods.Enabled = True
			Select Case lbModification.SelectedIndex
				Case CInt(Modifications.None)
				Case CInt(Modifications.CornerDetection)
					PopulateMethods(Modifications.CornerDetection)
				Case Else
			End Select

		End Sub

		' Populate the Modification Method list box.
		Private Sub PopulateMethods(ByVal modifications As Modifications)
			lbMethods.Items.Add("Susan Edge Detection") ' index 0
		End Sub

		Private Sub lbMethods_SelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs) Handles lbMethods.SelectedIndexChanged
			Select Case lbMethods.SelectedIndex
				Case CInt(Methods.none)
				Case CInt(Methods.Susan)

					DoSusan()
				Case Else
			End Select
		End Sub

		#Region "Susan Edge Detection"

		Private Sub DoSusan()
			SusanProperties = New SusanCornerProperties()
			SusanProperties.Dock = DockStyle.Fill
			ControlPanel.Controls.Add(SusanProperties)
		End Sub

		Private Sub s_ImageComplete(ByVal Corners As List(Of IntPoint))
			If processing IsNot Nothing Then
				If processing.InvokeRequired Then
					Dim d As New ImageCompleteDelegate(AddressOf s_ImageComplete)
					Me.Invoke(d, New Object() { Corners })
				Else
					processing.Close()
					processing = Nothing
				End If

			End If
			Log("Detection Finished")
			SetStatus(Corners.Count & " corners detected", True)
			DrawCorners(Corners)
		End Sub

		#End Region

		Private Sub DrawCorners(ByVal Corners As List(Of IntPoint))
			' Load image and create everything you need for drawing
			Dim image As New Bitmap(pbImage.Image)
			Dim graphics As Graphics = System.Drawing.Graphics.FromImage(image)
			Dim brush As New SolidBrush(Color.Red)
			Dim pen As New Pen(brush)


			' Visualization: Draw 3x3 boxes around the corners
			For Each corner As IntPoint In Corners
				graphics.DrawRectangle(pen, corner.X - 1, corner.Y - 1, 3, 3)
			Next corner

			' Display
			pbImage.Image = image
		End Sub

		Private Sub SetStatus(ByVal message As String, Optional ByVal Show As Boolean = True)

			If Show Then
				If statusStrip1.InvokeRequired Then
					Dim d As New StatusDelegate(AddressOf SetStatus)
					Me.Invoke(d, New Object() { message, Show })
				Else
					lblStatus.Text = message

				End If

			Else
				If statusStrip1.InvokeRequired Then
					Dim d As New StatusDelegate(AddressOf SetStatus)
					Me.Invoke(d, New Object() { message, Show })
				Else
					lblStatus.Text = message

				End If
			End If
		End Sub

		Private Sub btnReset_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btnReset.Click
			Select Case lbMethods.SelectedIndex
				Case CInt(Methods.none)
				Case CInt(Methods.Susan)
					Dim sp As SusanCornerProperties = TryCast(ControlPanel.Controls(0), SusanCornerProperties)
					sp.SetDefaults()

				Case Else
			End Select
		End Sub

		Private Sub btnProcess_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btnProcess.Click
			Select Case lbMethods.SelectedIndex
				Case CInt(Methods.none)
				Case CInt(Methods.Susan)
					Dim sp As SusanCornerProperties = TryCast(ControlPanel.Controls(0), SusanCornerProperties)

					processing = New FrmProcessing("Conducting Susan Edge Detection")
					processing.Show()
					Log("Conducting Susan Edge Detection")
					SetStatus("Please wait for edge detection")
					Dim s As New Susan(CurrentImage)
					AddHandler s.ImageComplete, AddressOf s_ImageComplete
					Dim t As New Task(Sub() s.GetCorners(sp.DifferenceThreshold,sp.GeometricalThreshold))
					t.Start()

				Case Else
			End Select
		End Sub



	End Class
End Namespace
